test/integration/storageversion/gc_test.go

```diff
@@ -0,0 +1,224 @@
+	encodingVersion = "storageversion.integration.test/v1"
+)
+
+func TestStorageVersionGarbageCollection(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.APIServerIdentity, true)()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StorageVersionAPI, true)()
+	result := kubeapiservertesting.StartTestServerOrDie(t, nil, nil, framework.SharedEtcd())
+	defer result.TearDownFn()
+
+	kubeclient, err := kubernetes.NewForConfig(result.ClientConfig)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	informers := informers.NewSharedInformerFactory(kubeclient, time.Second)
+	leaseInformer := informers.Coordination().V1().Leases()
+	storageVersionInformer := informers.Internal().V1alpha1().StorageVersions()
+
+	controller := storageversiongc.NewStorageVersionGC(kubeclient, leaseInformer, storageVersionInformer)
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	go leaseInformer.Informer().Run(stopCh)
+	go storageVersionInformer.Informer().Run(stopCh)
+	go controller.Run(stopCh)
+
+	t.Run("storage version with non-existing id should be GC'ed", func(t *testing.T) {
+		createTestStorageVersion(t, kubeclient, idNonExist)
+		assertStorageVersionGC(t, kubeclient)
+	})
+
+	createTestAPIServerIdentityLease(t, kubeclient, idA)
+	createTestAPIServerIdentityLease(t, kubeclient, idB)
+	t.Run("storage version with valid id should not be GC'ed", func(t *testing.T) {
+		createTestStorageVersion(t, kubeclient, idA)
+		if err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+			_, err := kubeclient.InternalV1alpha1().StorageVersions().Get(
+				context.TODO(), svName, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, nil
+		}); err == nil {
+			t.Fatalf("unexpected valid storage version getting GC'ed")
+		}
+		sv, err := kubeclient.InternalV1alpha1().StorageVersions().Get(
+			context.TODO(), svName, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("failed to retrieve valid storage version: %v", err)
+		}
+		if len(sv.Status.StorageVersions) != 1 {
+			t.Errorf("unexpected number of storage version entries, expected 1, got: %v",
+				sv.Status.StorageVersions)
+		}
+		expectedID := idA
+		if sv.Status.StorageVersions[0].APIServerID != expectedID {
+			t.Errorf("unexpected storage version entry id, expected %v, got: %v",
+				expectedID, sv.Status.StorageVersions[0].APIServerID)
+		}
+		if err := kubeclient.InternalV1alpha1().StorageVersions().Delete(
+			context.TODO(), svName, metav1.DeleteOptions{}); err != nil {
+			t.Fatalf("failed to cleanup valid storage version: %v", err)
+		}
+	})
+
+	t.Run("deleting an id should delete a storage version entry that it owns", func(t *testing.T) {
+		createTestStorageVersion(t, kubeclient, idA, idB)
+		assertStorageVersionEntries(t, kubeclient, 2, idA)
+		deleteTestAPIServerIdentityLease(t, kubeclient, idA)
```

Comment:
Can you also make apiserver A and apiserver B to have different encoding version, and then verify that after idA is deleted, the commonEncodingVersion is set?